### ISSUE TITLE
CTRP-style bounded quantification and `verifytypes`

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#9448]
### Discussed in https://github.com/microsoft/pyright/discussions/9445

<div type='discussions-op-text'>

<sup>Originally posted by **wence-** November 12, 2024</sup>
```
$ pyright --version
pyright 1.1.388
```

I have class hierarchy for a simple tree-structured DSL that uses CTRP-style generics to provide typing for the "children" of a node.

Distilled:

```python
from __future__ import annotations

from typing import Any

__all__ = [
    "Expr", "NonTerminal", "Terminal", "OneLevel",
]

class Node[T: Node[Any]]:
    children: tuple[T, ...]


class Expr(Node["Expr"]):
    pass

class Other(Node["Other"]):
    pass

class NonTerminal(Expr):
    def __init__(self, *exprs: Expr):
        self.children = exprs


class Terminal(Expr):
    def __init__(self):
        self.children = ()


class OneLevel(Expr):
    def __init__(self, *terminal: Terminal):
        self.children = terminal

# reveal_type(Expr().children)
# reveal_type(NonTerminal().children)
# reveal_type(Terminal().children)
# reveal_type(OneLevel().children)
```

If I ask pyright to `reveal_type` for an instance of any of `Expr`, `NonTerminal`, `OneLevel`, or `Terminal`, I obtain `tuple[Expr, ...]` (which I expect). So there are no type errors here.

Additionally, as expected, if I introduce:

```python
class OtherExpr(Node["OtherExpr"]):
   pass

class Bad(Expr):
   def __init__(self, *bad: OtherExpr):
       self.children = bad
```
Then pyright complains because `tuple[OtherExpr, ...]` is not a subtype of `tuple[Expr, ...]`.

However, if I install this as a package, call the above file `baseoverride.py`, and include:

```python
# __init__.py
from myexample.baseoverride import Node, Expr, NonTerminal, OneLevel

__all__ = ["Node", "Expr", "NonTerminal", "OneLevel"]
```

```toml
# pyproject.toml
[project]
name = "myexample"
requires-python = ">=3.12"
version = "1.0"
```

Then `pyright --verifytypes myexample` reports some errors:
```
$ pyright --verifytypes myexample
Module name: "myexample"
Package directory: "/home/coder/.conda/envs/rapids/lib/python3.12/site-packages/myexample"
Module directory: "/home/coder/.conda/envs/rapids/lib/python3.12/site-packages/myexample"
Path of py.typed file: "/home/coder/.conda/envs/rapids/lib/python3.12/site-packages/myexample/py.typed"

Public modules: 2
   myexample
   myexample.baseoverride

Symbols used in public interface:
myexample.baseoverride.NonTerminal.children
  /home/coder/.conda/envs/rapids/lib/python3.12/site-packages/myexample/baseoverride.py:21:14 - error: Ambiguous base class override
    Type declared in base class is "tuple[T@Node, ...]"
    Type inferred in child class is "tuple[Expr, ...]"
myexample.baseoverride.OneLevel.__init__
  /home/coder/.conda/envs/rapids/lib/python3.12/site-packages/myexample/baseoverride.py:30:9 - error: Type of parameter "terminal" is partially unknown
    Parameter type is "Terminal"
myexample.baseoverride.Terminal.children
  /home/coder/.conda/envs/rapids/lib/python3.12/site-packages/myexample/baseoverride.py:26:14 - error: Ambiguous base class override
    Type declared in base class is "tuple[T@Node, ...]"
    Type inferred in child class is "tuple[()]"
    Inferred child class type is missing type annotation and could be inferred differently by type checkers
myexample.baseoverride.OneLevel.children
  /home/coder/.conda/envs/rapids/lib/python3.12/site-packages/myexample/baseoverride.py:31:14 - error: Ambiguous base class override
    Type declared in base class is "tuple[T@Node, ...]"
    Type inferred in child class is "tuple[Terminal, ...]"

Symbols exported by "myexample": 17
  With known type: 8
  With ambiguous type: 9
  With unknown type: 0

Other symbols referenced but not exported by "myexample": 0
  With known type: 0
  With ambiguous type: 0
  With unknown type: 0

Symbols without documentation:
  Functions without docstring: 0
  Functions without default param: 0
  Classes without docstring: 10

Type completeness score: 47.1%

Completed in 0.552sec
```

I think these are all the same class of error, so let's consider the `NonTerminal` case:

```
myexample.baseoverride.NonTerminal.children
  /home/coder/.conda/envs/rapids/lib/python3.12/site-packages/myexample/baseoverride.py:21:14 - error: Ambiguous base class override
    Type declared in base class is "tuple[T@Node, ...]"
    Type inferred in child class is "tuple[Expr, ...]"
```

I think I agree that the type of the `children` property in `NonTerminal` is `tuple[Expr, ...]`. However, in the base class `Expr`, because the generic typevar `T` is bound (to `Expr`), I think the `children` property should _also_ have type `tuple[Expr, ...]`, whereas in `verifytypes` mode, pyright thinks it is `T@Node`. This is in contrast to the `reveal_type` result (where pyright thinks it is `tuple[Expr, ...]`).

For the other two cases, I suppose that the assignments to `children` _are_ to tighter types than `tuple[Expr, ...]`, but I think that this is OK, because `() is-a tuple[Expr, ...]` as is `tuple[Terminal, ...]`.

My rationale for having this setup is to avoid having to explicitly provide type annotations for the `children` property in every child class, since the only thing that I want to advertise of it is that is a `tuple[Expr, ...]`.
</div>

**Comment by wence-** [comment]
> The other two cases are ambiguous, and the type will be evaluated differently by pyright and mypy. If you want a consistent behavior across type checkers, you would need to provide an explicit type annotation here.

Can you explain this? 

If I understand correctly, you're saying that because (in `Terminal` say, but mutatis mutandis for `OneLevel`) I don't have an explicit `children: tuple[Expr, ...]` class-level type-declaration, then the type checker will attempt to infer the type of `self.children` based on the assignment in `__init__`.

That is (ignoring all the fancy generics)

```python
class Expr:
   children: tuple[Expr, ...]

class Terminal(Expr):
    def __init__(self):
        self.children = ()

reveal_type(Terminal().children) # => might be tuple[()], might be tuple[Expr, ...]?
```

But this I don't understand because I think I _did_ provide an explicit type annotation, on the base class in the hierarchy. So, in fact, I don't understand why pyright is trying to infer the type of the `children` attribute at all.

At the point of assignment, I think pyright knows that `self.children` must be a `tuple[Expr, ...]` because if I try and assign `(1, 2)` I definitely get a typing error.

**Comment by erictraut** [comment]
Pyright works the way you've described. Mypy does not in all cases, although its behavior has changed somewhat over the years. It used to always infer the type of instance variables in subclasses even if the parent class declared a wider type. My main point is that this behavior is not specified anywhere in the typing spec, so type checkers may not (and historically have not) had consistent behaviors here. If, as a library author, you want consistent behavior, you will need to be explicit. The job of the `--verifytypes` command is to help you identify such ambiguous cases.

**Comment by wence-** [comment]
Aha, thanks!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.391.
